### PR TITLE
Add three themes and theme switcher for Shoppy

### PR DESCRIPTION
**Pull request by heypoom**
Implements multiple themes support for Shoppy. Please see the demo video in the team channel.

### Screenshots

Stitch

CleanShot 2567-07-03 at 15 11 38@2x

Luminara

CleanShot 2567-07-03 at 15 11 53@2x

Pug & Play

CleanShot 2567-07-03 at 15 12 20@2x

### Sorry, this PR is huge!

Sadly, there isn't a nice way to break down this PR to make it easy to review. I've added review comments to make it easier for you. Note that most of them are CSS styles and configurations that you can glance over 🙏🏻

### CODE REVIEW

File: public/fonts/luminara.font.css [hunk 1]
```diff
@@ -0,0 +1,28 @@
+@import url('https://fonts.googleapis.com/css2?family=Arsenal:ital,wght@0,400;0,700;1,400;1,700&family=Libre+Caslon+Display&display=swap');
+
+@font-face {
+  font-family: 'Custom';
```
Comment (heypoom) [inline public/fonts/luminara.font.css:4]: This is because how our "Custom" font family works. We can remove this once we add proper fonts support, see https://github.com/metabase/metabase/issues/44394

File: index.html [hunk 2]
```diff
@@ -11,7 +11,7 @@
     <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
 
     <link
-      href="https://fonts.googleapis.com/css2?family=Barlow:wght@400;700&display=swap"
+      href="https://fonts.googleapis.com/css2?family=Lato:wght@400;700&display=swap"
```
Comment (heypoom) [inline index.html:14]: Used for the top theme switcher

File: api/drizzle/seeder/assign-images.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+import { db } from "../../src/utils/db"
+import { products as productSchema } from "../../src/schema/products"
+
+const SHOPS = ["stitch", "luminara", "pug"]
+
+/**
+ * Assign images to products based on their shops category.
+ */
+export async function assignImagesToProducts() {
```
Comment (heypoom) [inline api/drizzle/seeder/assign-images.ts:11]: One-time script used to assign images. Likely will never need to be ran again, but I added it here solely for future reference.

File: package.json [hunk 4]
```diff
@@ -65,6 +66,7 @@
   },
   "devDependencies": {
     "@babel/plugin-proposal-private-property-in-object": "^7.21.11",
+    "@iconify/react": "^5.0.1",
```
Comment (heypoom) [inline package.json:69]: The design from Alessio uses multiple iconsets, so `iconify/react` allows us to use multiple icon sets at once.

File: src/components/AppProvider.tsx [hunk 5]
```diff
@@ -26,42 +29,18 @@ const config: SDKConfig = {
...
-}
+export const AppProvider = ({ children }: Props) => {
+  const [themeKey] = useAtom($theme)
 
-export const AppProvider = ({ children }: Props) => (
-  <MetabaseProvider config={config} theme={theme}>
-    {children}
-  </MetabaseProvider>
-)
+  const theme = useMemo(() => {
```
Comment (heypoom) [inline src/components/AppProvider.tsx:35]: Metabase theme is loaded dynamically based on what theme we are selecting.

File: src/components/AuthCheck.tsx [hunk 6]
```diff
@@ -1,20 +1,44 @@
...
+  })
 
-  if (query.isLoading) return <FullPageLoader />
+  const loginMutation = useMutation({
+    mutationFn: loginToSite,
+    mutationKey: ["login"],
+  })
+
+  useEffect(() => {
+    if (!query.data?.email) {
```
Comment (heypoom) [inline src/components/AuthCheck.tsx:27]: If the user does not have a session yet, we login for them. This is needed for the JWT auth and SSO. We decided to simulate the login for them, so the UX is a bit easier than having to login.

File: src/components/RemountOnThemeChange.tsx [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import { useAtom } from "jotai"
+
+import { $theme } from "../store/theme"
+
+interface Props {
+  children: React.ReactNode
+}
+
+/**
+ * WORKAROUND: remount the children when the theme changes.
```
Comment (heypoom) [inline src/components/RemountOnThemeChange.tsx:11]: The MetabaseProvider has no way to know that the user session has changed, so this helps to avoid a full-page reload.

File: src/components/ThemeProvider.tsx [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    return createTheme(THEME_CONFIG_MAP[themeName].mantine)
+  }, [themeName])
+
+  // apply the CSS variables to the body, depending on the theme.
+  // this ensures the variables are applied in portals as well.
+  useEffect(() => {
+    document.body.setAttribute("data-theme", themeName)
+  }, [themeName])
+
+  return <MantineProvider theme={theme}>{props.children}</MantineProvider>
```
Comment (heypoom) [inline src/components/ThemeProvider.tsx:22]: This is for our own app's Mantine provider - not related to the SDK.

File: src/components/ThemeSwitcher.tsx [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+import { $theme } from "../store/theme"
+
+import { THEMES } from "../themes"
+import { loginToSite } from "../utils/switch-site"
+import { ThemeKey } from "../types/theme"
+
+export const ThemeSwitcher = () => {
+  const [activeTheme, setActiveTheme] = useAtom($theme)
+
+  const loginMutation = useMutation({
```
Comment (heypoom) [inline src/components/ThemeSwitcher.tsx:16]: This handles the loading state for us. We show a loading indicator on the site change button, so the user has an immediate feedback.

File: src/components/ThemeSwitcher.tsx [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+          {THEMES.map((theme) => {
+            const active = activeTheme === theme.key
+
+            return (
+              <Button
+                key={theme.key}
+                variant="outline"
+                size="xs"
+                color={active ? "#2B2F32" : "#7AC1FF"}
+                bg={active ? "#EAEAEA" : undefined}
```
Comment (heypoom) [inline src/components/ThemeSwitcher.tsx:54]: color is hard-coded here to avoid each of the 3 themes from tampering with the top-bar theme switcher (must remain the same)

File: src/components/layout/SidebarLinks.tsx [hunk 11]
```diff
@@ -40,11 +39,12 @@ const renderLink = (link: SidebarLink, child?: boolean) => (
...
           cx(
-            "hover:bg-transparent",
+            "hover:bg-transparent font-sans",
             props.className,
+            !child && "sidebar-link-root",
             child && "space-y-2",
-            child && !active && "text-light-grey",
-            active ? "text-primary" : "hover:text-gray-300",
+            child && !active && "sidebar-inactive-child",
+            active ? "dark-gradient" : "sidebar-active-child",
```
Comment (heypoom) [inline src/components/layout/SidebarLinks.tsx:47]: this is for themeability, as the sidebar differs a bit across themes.

File: src/hooks/useFontLoader.tsx [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+export function FontLoader() {
+  const [theme] = useAtom($theme)
+
+  return (
+    <div>
+      <link
+        type="text/css"
+        rel="stylesheet"
+        href={`/fonts/${theme}.font.css`}
```
Comment (heypoom) [inline src/hooks/useFontLoader.tsx:13]: we need a different set of Metabase font overrides

File: src/index.tsx [hunk 13]
```diff
@@ -1,26 +1,38 @@
...
 
 import "./styles/index.css"
 
+// CSS variables used for the demo app's themes.
+import "./themes/variables.css"
+
+// Theme overrides for each themes
+import "./themes/stitch.css"
+import "./themes/luminara.css"
+import "./themes/pug.css"
```
Comment (heypoom) [inline src/index.tsx:20]: these are scoped with `body[data-theme]`, so it's safe to import globally

File: src/routes/analytics/DashboardPage.tsx [hunk 14]
```diff
@@ -10,7 +12,13 @@ export function DashboardPage(props: Props) {
...
   return (
     <Box mih="100vh" className="dashboard-container smartscalar">
-      <InteractiveDashboard dashboardId={dashboardId} withTitle withDownloads />
+      <RemountOnThemeChange>
+        <InteractiveDashboard
+          dashboardId={dashboardId}
+          withTitle
+          withDownloads
+        />
+      </RemountOnThemeChange>
```
Comment (heypoom) [inline src/routes/analytics/DashboardPage.tsx:21]: Maintains the illusion of switching between themes, without forcing the entire page to go into a loading spinner.

File: src/routes/product-list/ProductCardFooter.tsx [hunk 15]
```diff
@@ -0,0 +1,26 @@
+import { Box, Text, Button, Divider } from "@mantine/core"
+import { useAtom } from "jotai"
+
+import { $theme } from "../../store/theme"
+
+export function ProductCardFooter() {
+  const [theme] = useAtom($theme)
+
+  if (theme === "luminara") {
```
Comment (heypoom) [inline src/routes/product-list/ProductCardFooter.tsx:9]: Luminara uses a divider instead of button.

File: src/store/theme.ts [hunk 16]
```diff
@@ -0,0 +1,6 @@
+import { atom } from "jotai"
+import { atomWithStorage } from "jotai/utils"
+
+import type { ThemeKey } from "../types/theme"
+
+export const $theme = atomWithStorage<ThemeKey>("theme", "stitch")
```
Comment (heypoom) [inline src/store/theme.ts:6]: Remembers what theme you last visited when you reload.

File: src/styles/index.css [hunk 17]
```diff
@@ -6,10 +6,17 @@
...
 #metabase-sdk-root {
   background: transparent;
 }
+
+body[data-theme="luminara"] {
+  background-image: url("/light-backdrop.png");
+}
+
+body:not([data-theme="luminara"]) {
+  background: var(--color-background) !important;
```
Comment (heypoom) [inline src/styles/index.css:21]: This is because of a bug in the SDK, specifically https://github.com/metabase/metabase/issues/44802. We delete this line once the bug has been fixed. TL;DR: EmbedFrame's style leaked.

File: src/styles/smartscalar-workaround.css [hunk 18]
```diff
@@ -2,30 +2,53 @@
...
-  color: #ADABA9;
+body[data-theme="luminara"] {
+  .smartscalar [data-testid="scalar-value"] {
+    font-weight: 200;
+    font-family: "Libre Caslon Display";
+  }
+}
+
+body[data-theme="stitch"] {
+  .smartscalar [data-card-key]>div:has(div[data-testid="scalar-container"]) {
```
Comment (heypoom) [inline src/styles/smartscalar-workaround.css:24]: Not pretty, but we have to use this workaround until the SDK officially supports fine-grained theming.

File: src/utils/switch-site.ts [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+import { logout } from "./logout"
+import { queryClient } from "./query-client"
+
+const SITE_TO_USER_MAP: Record<ThemeKey, string> = {
+  stitch: "rene@example.com",
+  luminara: "cecilia@example.com",
+  pug: "emily@example.com",
+}
+
+export async function loginToSite(_site: ThemeKey | null) {
```
Comment (heypoom) [inline src/utils/switch-site.ts:14]: Login to a new shop, so we get a sandboxed data.

File: src/hooks/useFontLoader.tsx [hunk 20]
```diff
@@ -0,0 +1,17 @@
+import { useAtom } from "jotai"
+
+import { $theme } from "../store/theme"
+
+export function FontLoader() {
+  const [theme] = useAtom($theme)
+
+  return (
+    <div>
```
Comment (WiNloSt) [inline src/hooks/useFontLoader.tsx:9]: Since it renders a `link` component, I don't think it needs to be nested under `div`. Like you could just return `link` directly.

File: src/hooks/useFontLoader.tsx
Comment (WiNloSt) [inline src/hooks/useFontLoader.tsx:1]: And this file seem to be a component, but the file name was still a hook, I think we should move this file to a proper location.

File: src/components/AuthCheck.tsx (hunk 6 above)
Comment (WiNloSt) [reply src/components/AuthCheck.tsx:27]: Can you point me to where the decision has been made? I'm curious.

File: src/components/AuthCheck.tsx [hunk 21]
```diff
@@ -1,20 +1,40 @@
-import { useQuery } from "@tanstack/react-query"
+import { useMutation, useQuery } from "@tanstack/react-query"
 import { Redirect } from "wouter"
 
 import { getUser } from "../utils/query-user"
 import { FullPageLoader } from "./Loader"
+import { loginToSite } from "../utils/switch-site"
+import { useEffect } from "react"
+import { ThemeKey } from "../types/theme"
+import { getCurrentTheme } from "../utils/current-theme"
```
Comment (WiNloSt) [inline src/components/AuthCheck.tsx:9]: these aren't used. I'd suggest you also add ESLint to catch unused imports. It could be another PR not to make this one too big unnecessarily.

File: src/utils/switch-site.ts (hunk 19 above)
Comment (WiNloSt) [inline src/utils/switch-site.ts:14]: nit: I'd just rename everything relating to `theme` to `site` instead. I feels weird to use `ThemeKey` type for `site` variable, but it feels ok to use `Site` (type) for `theme` variable don't you think?

File: src/utils/current-theme.ts [hunk 22]
```diff
@@ -0,0 +1,14 @@
+import { THEME_CONFIG_MAP } from "../themes"
+import { ThemeKey } from "../types/theme"
+
+const THEME_NAMES = Object.keys(THEME_CONFIG_MAP)
+
+export function getCurrentTheme(): ThemeKey {
+  const theme = JSON.parse(localStorage.getItem("theme")!)
```
Comment (WiNloSt) [inline src/utils/current-theme.ts:7]: I find it so disconnecting that we set this value to localStorge with `atomWithStorage`

https://github.com/metabase/shoppy/blob/5c4c99b4c650f18bfaebf229004f9e7c15e85dc3/src/store/theme.ts#L6

But we read it from localStorage directly here.

I'd kind of expect you'd try to read it from the atom somehow here, otherwise, this looks like a hack to me.

I found we could use `store` to [interact with atom outside React](https://jotai.org/docs/guides/using-store-outside-react). But I'm not totally sure how to apply that to this scenario, since I'm not familiar with Jotai.

File: src/components/RemountOnThemeChange.tsx [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * This causes the charts data to be re-fetched and re-rendered
+ * when the theme changes.
+ */
+export function RemountOnThemeChange(props: Props) {
```
Comment (WiNloSt) [inline src/components/RemountOnThemeChange.tsx:16]: This seems too overzealous in my opinion. I think we could achieve the same effect by forcing `key`.

```ts
export function RemountOnThemeChange(props: Props) {
  const [themeName] = useAtom($theme)
  return <Fragment key={themeName}>{props.children}</Fragment>
}
```

Could you see if this solves all your problems, I tried this with the dashboard page `admin/analytics/:id` and it reloaded the dashboard when changing themes

**Review by WiNloSt — COMMENTED** [review]
This is like 80% review. I'll continue later along with any fix you might incorporate.

File: src/components/AuthCheck.tsx (hunk 6 above)
Comment (heypoom) [reply src/components/AuthCheck.tsx:27]: @WiNloSt here you go: https://metaboat.slack.com/archives/C063Q3F1HPF/p1719509315975669?thread_ts=1719295515.987739&cid=C063Q3F1HPF

Comment (WiNloSt) [reply src/components/AuthCheck.tsx:27]: awesome thanks!

**Comment by heypoom** [comment]
Thanks for reviewing! Converting this PR to draft for now, so I can address both Alessio's design review feedback, and your code review feedback at once 🙏🏻

File: src/components/AppProvider.tsx [hunk 24]
```diff
@@ -12,6 +11,10 @@ import {
 
 import { MetabaseError, MetabaseLoader } from "./SdkStates"
 
+import { $theme } from "../store/theme"
```
[the remark this answers is not in the document]
Comment (heypoom) [reply src/components/AppProvider.tsx:14]: Good point, this was a convention from `nanostores` (a different library) rather than Jotai, let me update to `themeAtom`

File: src/hooks/useFontLoader.tsx (hunk 20 above)
Comment (heypoom) [reply src/hooks/useFontLoader.tsx]: Nice catch, I changed this file a lot so I forgot to clean it up 🙏🏻

File: src/hooks/useFontLoader.tsx
Comment (heypoom) [reply src/hooks/useFontLoader.tsx:1]: Whoops, nice catch!

File: src/utils/current-theme.ts (hunk 22 above)
Comment (heypoom) [reply src/utils/current-theme.ts]: Great catch! Replaced this with `store.get` 🙏🏻

File: src/routes/analytics/DashboardLinkCard.tsx [hunk 25]
```diff
@@ -12,20 +12,13 @@ export interface DashboardLinkCardProps {
...
+      <Card className="card gap-y-5 justify-between" h="100%" withBorder p={12}>
         <Box>
           <Title size="h4">{props.title}</Title>
 
-          <Text c="light-grey" fz="14px">
-            {props.description ?? "No description."}
-          </Text>
+          <Text fz="14px">{props.description ?? "No description."}</Text>
         </Box>
+        c="light-grey"
```
[the remark this answers is not in the document]
Comment (heypoom) [reply src/routes/analytics/DashboardLinkCard.tsx]: Thanks! Looks like my IDE glitched out...

File: src/styles/smartscalar-workaround.css (hunk 18 above)
Comment (heypoom) [reply src/styles/smartscalar-workaround.css:24]: @deniskaber I agree with this being misleading. Let me check with Alberto what he thinks. I think Alessio's goals for the 3 themes was to [show what would be eventually possible](https://metaboat.slack.com/archives/C063Q3F1HPF/p1719336477507209?thread_ts=1719295515.987739&cid=C063Q3F1HPF) with the SDK, so this might be a necessary evil.

> these, on purpose, are pushing the possibilities of what we could do (e.g., gradients, multiple fonts)

**Comment by heypoom** [comment]
NOTE: there is a pending [UI review feedback](https://www.notion.so/metabase/SDK-reference-app-Shoppy-themes-f2b539f2ea6445d8824d0caed66a6317?pvs=4#d817c03e4ad64f45a60627d30f205d60) from Alessio, but we can merge this first then address this in the next PR, so the changes wouldn't be larger than what it is.

File: src/store/theme.ts
Comment (WiNloSt) [inline src/store/theme.ts:1]: small thing here, but the file name is still `theme` wile all names are now `site` 🙏

File: src/components/SiteLogo.tsx [hunk 26]
```diff
@@ -0,0 +1,32 @@
+import { Text, Image } from "@mantine/core"
+import { useAtom } from "jotai"
+
+import { siteAtom } from "../store/theme"
+
+export function SiteLogo() {
+  const [theme] = useAtom(siteAtom)
```
Comment (WiNloSt) [inline src/components/SiteLogo.tsx:7]: nit: 
```suggestion
  const [site] = useAtom(siteAtom)
```

File: src/components/AppProvider.tsx [hunk 27]
```diff
@@ -26,42 +29,18 @@ const config: SDKConfig = {
...
-    scalar: {
-      value: {
-        fontSize: "47px",
-        lineHeight: "50px",
-      },
-    },
-  },
-}
+export const AppProvider = ({ children }: Props) => {
+  const [siteKey] = useAtom(siteAtom)
```
Comment (WiNloSt) [inline src/components/AppProvider.tsx:33]: I found we name this variable from `siteAtom` 3 ways
- `siteKey`
- `siteName`
- `theme`

If you like the explicitness I'd choose either of the first two, but I think `site` is also enough.

Can we ensure we name all of these variables the same? Or do you think it it's valid to name these differently based on where the atim is used?

File: src/themes/index.ts
Comment (WiNloSt) [inline src/themes/index.ts:1]: nit: the folder is still named `theme`, all other theme related files could stay here, but maybe this should be moved to somewhere else like `sites/*`

File: src/utils/current-site.ts
Comment (WiNloSt) [inline src/utils/current-site.ts:1]: Great job, this isn't hacky now 🙌

**Review by WiNloSt — APPROVED** [review]
I only have minor comments regarding file names and variable names, otherwise, this looks great. Also tested that it's working as expected. The force reloading dashboards still works when changing sites.

awesome job 👍

File: src/components/AppProvider.tsx (hunk 27 above)
Comment (heypoom) [reply src/components/AppProvider.tsx]: renamed - thanks for the keen eye!
